onnxscript/rewriter/fuse_relus_clips.py

```diff
@@ -0,0 +1,179 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Does the following transformation:
+- Relu(Relu(X)) -> Relu
+- Relu(Clip(X)) -> Clip
+- Clip(Relu(X)) -> Clip
+- Clip(Clip(X)) -> Clip
+"""
+
+import abc
+
+import numpy as np
+
+from onnxscript import ir
+from onnxscript.rewriter import pattern as orp
+
+
+class FuseSuccessiveRelu(orp.RewriteRuleClassBase):
+    """Replaces ``Relu(Relu(X))`` with ``Relu(X)``."""
+
+    def rewrite(self, op, x):
+        return op.Relu(x)
+
+    def pattern(self, op, x):
+        return op.Relu(op.Relu(x))
+
+
+class _FuseReluClipBase(orp.RewriteRuleClassBase, abc.ABC):
+    def rewrite(self, op, x, **kwargs):
+        first_clip_node = kwargs.get("out_first_clip").producer()
+        second_clip_node = None
+
+        if out_second_clip := kwargs.get("out_second_clip"):
+            second_clip_node = out_second_clip.producer()
+
+        min_clip, max_clip = self.compute_clip_min_max(first_clip_node, second_clip_node)
+        clip_initializers = []
+
+        if min_clip is not None:
+            clip_initializers.append(
+                op.initializer(min_clip, name=f"{first_clip_node.inputs[0].name}_min")
+            )
+
+        if max_clip is not None:
+            # ONNX Clip expects min and max inputs in order.
+            # If min is not provided, we insert None to maintain correct argument positions.
+            if min_clip is None:
+                clip_initializers.append(None)
+
+            clip_initializers.append(
+                op.initializer(max_clip, name=f"{first_clip_node.inputs[0].name}_max")
+            )
+
+        return op.Clip(x, *clip_initializers)
+
+    @abc.abstractmethod
+    def compute_clip_min_max(
+        self, first_clip_node: ir.Node, second_clip_node: ir.Node | None = None
+    ):
+        pass
+
+    def extract_min_max(self, node: ir.Node):
+        # Infer dtype from node first input
+        dtype = node.inputs[0].dtype.numpy()
+        min_clip, max_clip = None, None
+
+        if len(node.inputs) > 1:
+            min_input = node.inputs[1]
+            # If only a max is provided, min is implicitly None, so we check that
+            if min_input is not None:
+                min_clip = min_input.const_value.numpy()
+
+        if len(node.inputs) > 2:
+            max_clip = node.inputs[2].const_value.numpy()
+
+        return min_clip, max_clip, dtype
+
+    def check(self, context, **kwargs):
+        del context  # Unused
+        check_result = orp.MatchResult()
+
+        # check clip min/max are initializers
+        initializers = []
```

Comment:
nit: Using initializers as variable name seems a bit confusing here. Something like clip_min_max would be more understandable. Especially I think `is_graph_input()` and `is_initializer()` can express what you are doing here.